File: zZ/java/binary-tree-level-order-traversal-ii.java
/**
 * Binary Tree Level Order Traversal II 
 * 
 * Given a binary tree, return the bottom-up level order traversal of its nodes' values. (ie, from left to right, level by level from leaf to root).
 * 
 * For example:
 * Given binary tree {3,9,20,#,#,15,7},
 * 
 *     3
 *    / \
 *   9  20
 *     /  \
 *    15   7
 *  
 * 
 * return its bottom-up level order traversal as:
 * 
 * [
 *   [15,7],
 *   [9,20],
 *   [3]
 * ]
 */

import java.util.ArrayList;
import java.util.Collections;

public class Solution {
    public List<List<Integer>> levelOrderBottom(TreeNode root) {
        List<List<Integer>> result = new ArrayList<List<Integer>>();
        levelOrderBottomHelper(root, 0, result);
        Collections.reverse(result);
        return result;
    }

    private void levelOrderBottomHelper(TreeNode root, int level, List<List<Integer>> result) {
        if (root == null) {
            return;
        }

        if (result.size() == level) {
            result.add(new ArrayList<Ingeger>());
        }
        result.get(level).add(root.val);

        levelOrderBottomHelper(root.left, level + 1, result);
        levelOrderBottomHelper(root.right, level + 1, result);
    }
}

public class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode(int x) { val = x; }
}

File: data-structure-and-algorithm/typescript/leetcode/src/binary-tree-level-order-traversal-ii.ts
/*
Given the root of a binary tree, return the bottom-up level order traversal of its nodes' values. (i.e., from left to right, level by level from leaf to root).



Example 1:


Input: root = [3,9,20,null,null,15,7]
Output: [[15,7],[9,20],[3]]
Example 2:

Input: root = [1]
Output: [[1]]
Example 3:

Input: root = []
Output: []


Constraints:

The number of nodes in the tree is in the range [0, 2000].
-1000 <= Node.val <= 1000
*/

import {Nullable} from './util/object';
import {TreeNode} from './data-structure/BinaryTree';

function levelOrderBottom(root: Nullable<TreeNode>): number[][] {
  const result: number[][] = [];
  levelOrderDfs(root, 0);
  return result.reverse();

  function levelOrderDfs(node: Nullable<TreeNode>, level: number) {
    if (node === null) {
      return;
    }
    if (result.length === level) {
      result.push([]);
    }
    result[level].push(node.data);
    levelOrderDfs(node.left, level + 1);
    levelOrderDfs(node.right, level + 1);
  }
}
